### PR TITLE
Stochastic Depth Curriculum: progressive block dropping during early training

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The 3-layer Transolver trains all blocks simultaneously from epoch 0. If later blocks produce noisy gradients early in training (before they learn useful transformations), they corrupt the representations that earlier blocks are trying to learn. This is the **early-layer bottleneck** problem in deep networks.

**Stochastic depth** (Huang et al., ECCV 2016) randomly drops entire blocks during training with a per-block probability. The **curriculum** variant decays the drop probability to 0 over the first half of training:
- Epochs 0-80: block 2 has drop probability decaying from 0.3→0, block 1 from 0.15→0
- Epochs 80+: all blocks always active (standard inference behavior)
- Block 0 (first): never dropped

This forces blocks 0-1 to develop **independently robust representations** before block 2 is reliably active. It's a structural regularizer — much stronger than activation dropout.

**Why it might help OOD:** The model's failure to generalize tandem pressure to NACA6416 likely originates in the early blocks where geometry features are extracted. Stochastic depth pressure-tests these blocks to be self-sufficient, which regularizes against late-block overfitting to in-distribution patterns.

**Zero inference cost.** Drop probability is 0 after epoch 80 and always 0 at eval time. torch.compile compatible (the conditional skip is a simple if statement).

**Expected improvement:** -1 to -3% across all metrics from better regularization. Largest expected gains on OOD splits (p_oodc, p_tan, p_re).

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flags

```python
stochastic_depth: bool = False       # progressive block dropping during training
stochastic_depth_max_drop: float = 0.3  # max drop prob for the last block (decays to 0)
stochastic_depth_warmup: int = 80    # epochs until drop prob reaches 0
```

### Step 2: Modify TransolverBlock to accept drop probability

In `TransolverBlock.forward`, add a `drop_prob` parameter:

```python
def forward(self, fx, x, T, B, N, ..., drop_prob=0.0):
    # Stochastic depth: skip entire block with probability drop_prob
    if self.training and drop_prob > 0.0:
        if torch.rand(1, device=fx.device).item() < drop_prob:
            return fx, fx_deep if ... else fx  # pass through unchanged
    
    # ... normal forward pass ...
```

**Important:** When the block is dropped, return `fx` unchanged (identity). The `fx_deep` output (for pressure_deep) should also be `fx` unchanged.

### Step 3: Compute per-block drop probabilities in training loop

At the start of each epoch:
```python
if cfg.stochastic_depth and epoch < cfg.stochastic_depth_warmup:
    # Linear decay from max_drop to 0 over warmup epochs
    decay_factor = 1.0 - (epoch / cfg.stochastic_depth_warmup)
    # Per-block drop prob: linear scaling, block 0 = 0, last block = max
    n_blocks = cfg.n_layers
    drop_probs = [
        cfg.stochastic_depth_max_drop * (i / (n_blocks - 1)) * decay_factor
        for i in range(n_blocks)
    ]
    # drop_probs for 3 blocks: [0.0, 0.15*decay, 0.30*decay]
else:
    drop_probs = [0.0] * cfg.n_layers
```

Pass `drop_probs[i]` to each block's forward call.

### Step 4: Log drop statistics

```python
if cfg.stochastic_depth and step % 50 == 0:
    wandb.log({f'stochastic_depth/drop_prob_block{i}': dp for i, dp in enumerate(drop_probs)})
```

### Step 5: Ensure eval loops DON'T drop

In all validation/eval loops, pass `drop_prob=0.0` to all blocks (or simply don't pass the parameter — default is 0.0).

### Step 6: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent thorfinn --wandb_name "thorfinn/stochastic-depth-s42" \
  --wandb_group "round17/stochastic-depth-curriculum" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --stochastic_depth --stochastic_depth_max_drop 0.3 --stochastic_depth_warmup 80

# Seed 73 — identical but --seed 73 --wandb_name "thorfinn/stochastic-depth-s73"
```

### Step 7: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, comparison to baseline, W&B run IDs. Report what epochs the drop probs reached 0 and whether training dynamics differed from baseline.

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: hgml7i2r (seed 42), qic03vrg (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent thorfinn --wandb_name "thorfinn/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Stochastic Depth Curriculum — Negative Result

**Summary:** Stochastic depth curriculum significantly **regresses** all metrics vs baseline. The regularization effect of randomly dropping blocks during early training hurts rather than helps this model.

### Metrics (Surface Pressure MAE, Pa)

| Metric | Baseline (2-seed avg) | SD s42 | SD s73 | SD 2-seed avg | Delta |
|--------|----------------------|--------|--------|---------------|-------|
| p_in   | **11.979** | 16.9 | 15.1 | 16.000 | **+33.6%** |
| p_tan  | **28.341** | 30.4 | 29.7 | 30.050 | **+6.0%** |
| p_oodc | **7.643**  | 9.9  | 8.9  | 9.400  | **+23.0%** |
| p_re   | **6.300**  | 8.3  | 7.5  | 7.900  | **+25.4%** |

### Val Loss

| Split | Baseline s42 | SD s42 | Baseline s73 | SD s73 |
|-------|-------------|--------|-------------|--------|
| val_in_dist | - | 0.3374 | - | 0.2826 |
| val_tandem_transfer | - | 0.9383 | - | 0.9047 |
| val_ood_cond | - | 0.2523 | - | 0.2250 |
| val_ood_re | - | 0.2418 | - | 0.2208 |

### Experiment Details

- **Seed 42:** W&B run `swddfzc7`, 111 epochs, 180.9 min, 45.1GB VRAM
- **Seed 73:** W&B run `t1aww5gn`, 128 epochs, 180.3 min, 42.3GB VRAM
- **Config:** `--stochastic_depth --stochastic_depth_max_drop 0.3 --stochastic_depth_warmup 80`
- Block 0: never dropped. Block 1: max 0.15 at epoch 0, decaying linearly to 0 by epoch 80. Block 2 (last): receives nonzero drop_prob but skips via `not self.last_layer` guard (output block cannot be dropped).

### Command Used

```bash
cd cfd_tandemfoil && python train.py \
  --agent thorfinn --wandb_name "thorfinn/stochastic-depth-s42" \
  --wandb_group "round17/stochastic-depth-curriculum" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --stochastic_depth --stochastic_depth_max_drop 0.3 --stochastic_depth_warmup 80
```

### What Happened

The stochastic depth curriculum **hurt all metrics substantially**, with the largest regressions on in-distribution (+33.6%) and OOD-condition (+23.0%) pressure.

**Why it failed:** With only 3 Transolver blocks, randomly skipping block 1 during training is too disruptive. In deeper networks (ResNet-110, 1000+ layers), stochastic depth works because there are many redundant blocks that learn overlapping representations. With 3 blocks, each block is **load-bearing** — block 0 handles spatial encoding, block 1 does cross-slice attention refinement, and block 2 produces the final output. Dropping block 1 forces the model to learn a degenerate shortcut where block 0 → block 2 must work directly, which conflicts with the normal 3-block pipeline.

Additionally, the `pressure_deep` pathway takes intermediate features from blocks — disrupting block 1 during training creates inconsistent gradient signals through the deep pressure branch, further degrading learning.

The model did still converge to reasonable absolute numbers, but significantly worse than the well-tuned baseline. The training curves show no benefit at any point during training — the regularization was harmful from epoch 0.

### Suggested Follow-ups

1. **Not recommended:** Reducing `stochastic_depth_max_drop` to 0.1 is unlikely to help — the fundamental issue is that 3 blocks is too shallow for block-level dropout.
2. **Alternative regularization:** If regularization is the goal, attention-level dropout within blocks (DropPath on the attention output, not the entire block) would be less disruptive. But the baseline already has good regularization from augmentations and EMA.
3. **Focus elsewhere:** The model's tandem transfer gap is more likely an architecture/conditioning issue than a regularization issue. Approaches that improve information flow (e.g., backbone AdaLN conditioning, iterative refinement) are more promising directions.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2230 because ADVISOR: Closing — all metrics regressed significantly (p_in +33.6%, p_oodc +23.0%, p_re +25.4%, p_tan +6.0%).

Your analysis is exactly right: with only 3 blocks, each is load-bearing. Stochastic depth needs deep redundant networks (100+ layers) where dropping a block doesn't destroy the information pathway. In our 3-block Transolver, block 1 is critical for cross-slice attention refinement, and the pressure_deep intermediate branch creates additional gradient dependencies on block 1.

The reduced epoch count (111-128 vs 145-155 baseline) from the overhead also hurts — less time for the model to recover from early-training disruption.

Good observation about DropPath as a less disruptive alternative. Moving you to a different regularization approach..
